### ISSUE TITLE
condAll

### ISSUE THREAD

**Issue by Undistraction** [char0n/ramda-adjunct#359]
A version of Ramda's [cond](http://ramdajs.com/docs/#cond) where every predicate is run with the initial value transformed along the way by functions for which the predicate returned true.

```
const condAll = curry((a, value) => reduce(
  (acc, [predicate, f]) => ifElse(
    predicate,
    f,
    identity
  )(acc),
  value,
  a
))
```
[REPL](http://ramdajs.com/repl/?v=0.25.0#?const%20isAlpha%20%3D%20T%0Aconst%20isBravo%20%3D%20F%0Aconst%20isCharlie%20%3D%20T%0Aconst%20isDelta%20%3D%20F%0Aconst%20isEcho%20%3D%20F%0A%0Aconst%20alpha%20%3D%20append%28%27a%27%29%0Aconst%20bravo%20%3D%20append%28%27b%27%29%0Aconst%20charlie%20%3D%20append%28%27c%27%29%0Aconst%20delta%20%3D%20append%28%27d%27%29%0Aconst%20echo%20%3D%20append%28%27e%27%29%0A%0Aconst%20condAll%20%3D%20curry%28%28a%2C%20value%29%20%3D%3E%20reduce%28%0A%20%20%28acc%2C%20%5Bpredicate%2C%20f%5D%29%20%3D%3E%20ifElse%28%0A%20%20%20%20predicate%2C%0A%20%20%20%20f%2C%0A%20%20%20%20identity%0A%20%20%29%28acc%29%2C%0A%20%20value%2C%0A%20%20a%0A%29%29%0A%0Aconst%20condAllImp%20%3D%20condAll%28%5B%0A%20%20%5BisAlpha%2C%20alpha%5D%2C%0A%20%20%5BisBravo%2C%20bravo%5D%2C%0A%20%20%5BisCharlie%2C%20charlie%5D%2C%0A%20%20%5BisDelta%2C%20delta%5D%2C%0A%20%20%5BisEcho%2C%20echo%5D%2C%0A%5D%29%0A%0AcondAllImp%28%5B%5D%29%20%2F%2F%20%5Ba%2C%20c%5D)

**Comment by guillaumearm** [comment]
Love this one, it will replace all my `when pipelines` ;-)

```js
pipe(
  when(predicate, doThis),
  when(predicate2, doThat),
  ...
)
```

**Comment by Undistraction** [comment]
> Love this one, it will replace all my when pipelines ;-)

Exactly. I often find myself processing a value based on a set of arguments to the function (or a config argument), taking the value through a series of transformations based on the arg values. I think this is a nice way to do it, though perhaps there is a better way already in existence.

**Comment by char0n** [comment]
Interesting idea, but I have point to discusss.

**The case of declarativness**

```js
condAll([
  [isAlpha, alpha],
  [isBravo, bravo],
  [isCharlie, charlie],
  [isDelta, delta],
  [isEcho, echo],
])
```

```js
pipe(
  when(isAlpha, alpha),
  when(isBravo, bravo),
  when(isCharlie, charlie),
  when( isDelta, delta),
  when(isEcho, echo),
),
```

Is the case for `condAll` strong enough to have it in RA ?
Which version is more declarative ? (IMHO I think the second one)


**The case of flexibility**

The second version allows to inject other HOFs

```js
pipe(
  when(isAlpha, alpha),
  when(isBravo, bravo),
  ifElse(
    isAlpha,
    alpha,
    bravo,
  ),
  when(isCharlie, charlie),
  when( isDelta, delta),
  when(isEcho, echo),
),
```

I still like the idea of `condAll` bet let's first discuss the questions I have.

**Comment by Undistraction** [comment]
I would argue that both have a place, and I would choose whichever was most readable for the situation. 

I would see myself using `condAll` in specific situations where all I want to do is effectively `when(predicate, f)` for a number of values, and in such a case I think there is less to type and less visual noise. There is also less mental parsing - if you understand what the function does, you don't need to extrapolate further, whereas with the `pipe` version, you will need to check if it does anything else. 

 I wonder if `condAll` is  a little vague given it is effectively a chain of `when`s, so perhaps a name based around `when` instead of `cond` would make more sense here. Using `pipe` is undoubtedly more flexible, but I see this as a specialisation for a particular case.

**Comment by char0n** [comment]
Yes, thought about it some more and I tend to agree. condAll is a good
name, 'when' is only an implementation detail. You could easily use ifElse
or any other construct (as you did in your composition).
